README.md
![Build status](https://github.com/avdata99/yaasr/workflows/Build/badge.svg?branch=main)
[![Pypi py version](https://img.shields.io/pypi/pyversions/yaasr)](https://pypi.org/project/yaasr/)
[![GitHub All Releases](https://img.shields.io/github/downloads/avdata99/yaasr/total)](https://github.com/avdata99/yaasr/releases)
[![GitHub Issues](https://img.shields.io/github/issues/avdata99/yaasr)](https://github.com/avdata99/yaasr/issues)
[![GitHub PR](https://img.shields.io/github/issues-pr/avdata99/yaasr)](https://github.com/avdata99/yaasr/pulls)
[![Licence](https://img.shields.io/github/license/avdata99/yaasr)](https://github.com/avdata99/yaasr/blob/main/LICENSE)
[![Last Commit](https://img.shields.io/github/last-commit/avdata99/yaasr)](https://github.com/avdata99/yaasr/commits/main)

# Yet Another Audio Stream Recorder

Audio stream recorded and static DB for radio stations.  
**Why?** Because other tools are (probably, [is not clear](https://github.com/mgaitan/radiocut_downloader/pull/18#issuecomment-526191324)) not availabe to use.

## Install

Install with pip

```
pip install yaasr
```

## Usage

### From Python

#### Save predefined audio locally

Load a pre-defined stream and save 5 audio chunks of 60 seconds

```python
from yaasr.recorder.stream import YStream

ys = YStream('radio-universidad-cordoba-argentina')
ys.load()
ys.record(total_seconds=300, chunk_bytes_size=1024, chunk_time_size=60)
```

You will see new audio files at your local folder

#### Save custom audio locally

Load a custom stream and save 5 audio chunks of 60 seconds

```python
from yaasr.recorder.stream import YStream

""" test custom stream """
ys = YStream(stream_name='my-custom-stream')
ys.destination_folder = 'some-path/my-chunks-folder'
ys.title = 'My radio title'
ys.short_name = 'my-radio'

# list of stream (if first fails the second should be used)
ys.streams = [
    {'url': 'https://my-radio.org/stream'}
]

ys.record(total_seconds=300, chunk_bytes_size=1024, chunk_time_size=5)
```

You will see new audio files at `some-path/my-chunks-folder` folder

#### Upload to Google Cloud Storage

```python
from yaasr.recorder.stream import YStream
from yaasr.processors.audio.reduce import reformat
from yaasr.processors.archive.google_cloud import upload_to_google_cloud_storage

# os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = "google-cloud-storage-credential.json"
ys = YStream('radio-universidad-cordoba-argentina')
ys.load()

# post-processors (you can combine or create new processors)
ys.post_process_functions = [
    {
        'fn': reformat,
        'params': {
            'audio_format': 'mp3',
            'mono': True,
            'delete_on_success': True
        }
    },
    {
        'fn': upload_to_google_cloud_storage,
        'params': {
            'bucket_name': 'parlarispa-radio',
            'delete_on_success': True
        }
    }
]
ys.record(total_seconds=300, chunk_bytes_size=1024, chunk_time_size=60)
```

![google-cloud-storage-list](docs/img/google-cloud-storage-list.png)
![google-cloud-storage-element](docs/img/google-cloud-storage-element.png)

#### Upload audio chunks using ssh

Post process audio to MP3 16Khz and upload via ssh the result cleaning local files after the process

```python
from yaasr.recorder.stream import YStream
from yaasr.processors.audio.reduce import reformat
from yaasr.processors.archive.ssh import upload_ssh


ys = YStream('radio-universidad-cordoba-argentina')
ys.load()

# post-processors (you can combine or create new processors)
ys.post_process_functions = [
    {
        'fn': reformat,
        'params': {
            'audio_format': 'mp3',
            'delete_on_success': True
        }
    },
    {
        'fn': upload_ssh,
        'params': {
            'host': 'myhost.com',
            'user': 'username',
            'password': 'mypass',
            'destination_folder': '/home/username/audios/',
            'port': 901,
            'delete_on_success': True
        }
    }
]
ys.record(total_seconds=300, chunk_bytes_size=1024, chunk_time_size=60)
```

![ssh files](docs/img/sshed.png)

### Usage from command line

List all available streams

```
$ yaasr ls
radio-bio-bio-santiago-chile: https://unlimited4-us.dps.live/biobiosantiago/aac/icecast.audio
radio-universidad-cordoba-argentina: https://sp4.colombiatelecom.com.co:10995/stream
```

Info about a stream

```
$ yaasr info --stream radio-bio-bio-santiago-chile
{
    "title": "Bio Bio Santiago de Chile",
    "web": "https://vivo.biobiochile.cl/player/",
    "streams": [
        {
            "url": "https://unlimited4-us.dps.live/biobiosantiago/aac/icecast.audio",
            "extension": "aac"
        }
    ]
}
```

Record a stream

```
$ yaasr record \
    --stream radio-bio-bio-santiago-chile \
    --total_seconds 90 \
    --chunk_bytes_size 512 \
    --chunk_time_size 30

2021-02-14 18:27:20,382 - yaasr.recorder.stream - INFO - Attempt to record from https://unlimited4-us.dps.live/biobiosantiago/aac/icecast.audio
2021-02-14 18:27:21,244 - yaasr.recorder.stream - INFO - Recording from https://unlimited4-us.dps.live/biobiosantiago/aac/icecast.audio
2021-02-14 18:27:56,923 - yaasr.recorder.stream - INFO - 2021-02-14 18:27:56.923239 Elapsed 0:00:35.679521 Finish chunk 1274
2021-02-14 18:27:56,924 - yaasr.recorder.stream - INFO - Chunk finished
2021-02-14 18:28:27,132 - yaasr.recorder.stream - INFO - 2021-02-14 18:28:27.131768 Elapsed 0:01:05.888050 Finish chunk 1981
2021-02-14 18:28:27,132 - yaasr.recorder.stream - INFO - Chunk finished
2021-02-14 18:28:51,294 - yaasr.recorder.stream - INFO - Finish recording 2021-02-14 18:28:51.294881
2021-02-14 18:28:51,295 - yaasr.recorder.stream - INFO - Chunk finished
```

### More docs

 - Add recording task to [supervisor](docs/run-with-supervisor.md)


CHANGELOG
# Changelog

## [0.4.11] 2021-04-03

### Notify external apps

Allow adding external URLs and params to notify recording tasks

## [0.4.0] 2021-02-16

### Supervidor
 - Allow creating supervisor tasks to save 24/7 all radio stations

## [0.3.9] 2021-02-16

### File names
 - Allow to use a short name to improve chunk names

## [0.3.8] 2021-02-16

### Google Cloud storage
 - Allow to run it from command line

## [0.3.4] 2021-02-15

### Google Cloud storage
 - Added Google Cloud Storage Archiving feature
 - Allow reduce Audio as mono

## [0.3.3] 2021-02-14

### Command line tools added
 - `yaasr ls`
 - `yaasr test`
 - `yaasr info`
 - `yaasr record`

## [0.3.2] 2021-02-14

### Post processing audios
Allow post-process functions:

 - Allow to reformat audios (e.g. mp3 bitrate 16K)
 - Allow upload final audios using SSH
 - Allow delete previous version file after post-process

## [0.2.1] 2021-02-13

### Added
- Allow to save from external URL


yaasr/terminal/__init__.py
"""
Iterate over priority domains and send data to server
"""
import argparse
import json
import logging.config
import os
from yaasr import get_all_streams
from yaasr.recorder.stream import YStream
from yaasr.processors.audio.reduce import reformat
from yaasr.processors.archive.google_cloud import upload_to_google_cloud_storage
from yaasr.terminal.supervisor import setup_supervisor


def ls(test_streams=False):
    """ List all available streams """
    # list all streams
    streams = get_all_streams()
    for stream in streams:
        ys = YStream(stream)
        ys.load()
        print(f'{stream}: {ys.streams[0]["url"]}')
        if test_streams:
            ys.record(total_seconds=30, chunk_bytes_size=512, chunk_time_size=1000)


def info(stream):
    """ Shows stream info """
    # list all streams
    ys = YStream(stream)
    ys.load()
    data = json.dumps(ys.data, indent=4)
    print(data)


def record(stream, total_seconds=300, chunk_bytes_size=256, chunk_time_size=60):
    """ Shows stream info """
    # list all streams
    ys = YStream(stream)
    ys.load()
    ys.record(total_seconds=total_seconds, chunk_bytes_size=chunk_bytes_size, chunk_time_size=chunk_time_size)


def compress_and_google_store(stream,
                              bucket_name,
                              total_seconds=300,
                              chunk_bytes_size=256,
                              chunk_time_size=60,
                              audio_format='ogg'):
    """ Shows stream info """

    ys = YStream(stream)
    ys.load()
    ys.post_process_functions = [
        {
            'fn': reformat,
            'params': {
                'audio_format': audio_format,
                'mono': True,
                'delete_on_success': True
            }
        },
        {
            'fn': upload_to_google_cloud_storage,
            'params': {
                'bucket_name': bucket_name,
                'delete_on_success': True
            }
        }
    ]

    ys.record(total_seconds=total_seconds, chunk_bytes_size=chunk_bytes_size, chunk_time_size=chunk_time_size)


def main():
    # TODO fix path or replace to local settings
    logging.config.fileConfig('yaasr/log.conf')
    parser = argparse.ArgumentParser()
    parser.add_argument('command', help='Command to run')
    parser.add_argument('--log_level', nargs='?', default='INFO', type=str)
    parser.add_argument('--stream', nargs='?', default=None, type=str)
    # record parameters
    parser.add_argument('--total_seconds', nargs='?', default=0, type=int)
    parser.add_argument('--chunk_bytes_size', nargs='?', default=256, type=int)
    parser.add_argument('--chunk_time_size', nargs='?', default=1200, type=int)
    # compress parameters
    parser.add_argument('--audio_format', nargs='?', default='mp3', choices=['mp3', 'ogg'], type=str)
    parser.add_argument('--bucket_name', nargs='?', default=None, type=str)

    # credentials
    parser.add_argument('--google-credentials', nargs='?', default=None, type=str)
    parser.add_argument('--system-user', nargs='?', default=None, type=str)

    args = parser.parse_args()

    if args.google_credentials is not None:
        os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = args.google_credentials

    if args.command == 'ls':
        return ls()
    elif args.command == 'test':
        return ls(test_stream=True)
    elif args.command == 'info':
        return info(stream=args.stream)
    elif args.command == 'record':
        return record(
            stream=args.stream,
            total_seconds=args.total_seconds,
            chunk_bytes_size=args.chunk_bytes_size,
            chunk_time_size=args.chunk_time_size
        )
    elif args.command == 'compress-and-google-store':
        if args.bucket_name is None:
            raise Exception('bucket name is required')

        return compress_and_google_store(
            stream=args.stream,
            bucket_name=args.bucket_name,
            total_seconds=args.total_seconds,
            chunk_bytes_size=args.chunk_bytes_size,
            chunk_time_size=args.chunk_time_size,
            audio_format=args.audio_format
        )
    elif args.command == 'supervisor':
        return setup_supervisor(
            stream_name=args.stream,
            bucket_name=args.bucket_name,
            system_user_name=args.system_user,
            google_credentials_path=args.google_credentials,
            total_seconds=args.total_seconds,
            chunk_bytes_size=args.chunk_bytes_size,
            chunk_time_size=args.chunk_time_size
        )
